map-storage/src/Services/CustomEntityCollectionService.ts

```diff
@@ -0,0 +1,113 @@
+                ...entityToModify,
+                name,
+                tags,
+                depthOffset,
+                collisionGrid,
+            };
+            await fileSystem.writeStringAsFile(
+                this.getEntityCollectionFileVirtualPath(),
+                JSON.stringify(customEntityCollection)
+            );
+        }
+    }
+
+    public async deleteEntity(deleteCustomEntityMessage: DeleteCustomEntityMessage) {
+        const { id } = deleteCustomEntityMessage;
+        const customEntityCollectionFileContent = await this.readOrCreateEntitiesCollectionFile();
+        const customEntityCollection = EntityCollectionRaw.parse(JSON.parse(customEntityCollectionFileContent));
+        const customEntityToDelete = customEntityCollection.collection.find((entity) => entity.id === id);
+        customEntityCollection.collection = customEntityCollection.collection.filter(
+            (customEntity) => customEntity.id !== id
+        );
+        await fileSystem.writeStringAsFile(
+            this.getEntityCollectionFileVirtualPath(),
+            JSON.stringify(customEntityCollection)
+        );
+        if (customEntityToDelete) {
+            await fileSystem.deleteFiles(this.getEntityToUploadVirtualPath(customEntityToDelete.imagePath));
+        }
+    }
+
+    private async readOrCreateEntitiesCollectionFile() {
+        const entityCollectionFileVirtualPath = this.getEntityCollectionFileVirtualPath();
+        const fileExist = await fileSystem.exist(entityCollectionFileVirtualPath);
+        if (!fileExist) {
+            const entityCollectionFile: EntityCollectionRaw = {
+                version: "1.0",
+                collection: [],
+                collectionName: "custom entities",
+                tags: [],
+            };
+            await fileSystem.writeStringAsFile(entityCollectionFileVirtualPath, JSON.stringify(entityCollectionFile));
+        }
+        //Check current version and migrate to new one
+        return fileSystem.readFileAsString(entityCollectionFileVirtualPath);
+    }
+
+    private mapEntityFromUploadEntityMessageToEntityRawPrefab(
+        uploadEntityMessage: UploadEntityMessage
+    ): EntityRawPrefab {
+        return EntityRawPrefab.parse({
+            ...uploadEntityMessage,
+            direction: mapCustomEntityDirectionToDirection(uploadEntityMessage.direction),
+        });
+    }
+
+    private async addEntityInEntityCollectionFile(entityToAddInCollection: EntityRawPrefab) {
+        const customEntityCollectionFileContent = await this.readOrCreateEntitiesCollectionFile();
+        const customEntityCollection = EntityCollectionRaw.parse(JSON.parse(customEntityCollectionFileContent));
+        customEntityCollection.collection.push(entityToAddInCollection);
+        await fileSystem.writeStringAsFile(
+            this.getEntityCollectionFileVirtualPath(),
+            JSON.stringify(customEntityCollection)
+        );
```

Comment:
addEntity and deleteEntity should share a common lock as we might have a race condition here.

You can implement that with a simple promise:

```
export class CustomEntityCollectionService {
   private lock: Promise<void>;

  private async addEntityInEntityCollectionFile(entityToAddInCollection: EntityRawPrefab) {
    this.lock = this.lock.then(() => {
      //...
    });
  }

  // Same thing for delete
  }
```